### ISSUE TITLE
Choose client-side data storage

### ISSUE THREAD

**Issue by biancadanforth** [mozilla/lightbeam-we#13]
In our MVP web extension, we use the _storage_ API, but it may not suit our needs long term (e.g. for advanced filtering). We should figure out what our storage requirements are and what the most appropriate option will be (other options: indexedDB, SQLite...).

**Comment by jonathanKingston** [comment]
This is post MVP as we are using the browser storage for now which is good for the MVP at least.

**Comment by biancadanforth** [comment]
Frederik Braun (aka Freddy from Paul's team) recommends [localForage.js](https://localforage.github.io/localForage/).

**Comment by biancadanforth** [comment]
Here are my [results](https://docs.google.com/document/d/1AXfXcDF_mdr7e30nWVUgS_m5kUqMvVDlNZdOHqiCOhQ/edit?usp=sharing) from some basic tests comparing localforage vs Dexie.

TL;DR: localforage seems better.

Please let me know what you guys think. :)

**Comment by jonathanKingston** [comment]
I added an issue on your repo to improve the tests. Hopefully those changes
will make it more realistic but local forage is super likely still to win
too.

**Comment by biancadanforth** [comment]
[JSPerf](jsperf.com) test results with N = 100^ websites (units: Ops/s):
* `clearAllDatabaseFromBefore` (localforage: ~3842 - 36% faster, Dexie: ~2600)
  * Not a method we will call frequently; and Dexie's implementation is actually the IndexedDB implementation, as I had trouble getting Dexie's to work.
* `setAll` (localforage: ~3866, Dexie: ~4270 - 9% faster)
* `getAll` (localforage: ~3502, Dexie: ~3917 - 10% faster)
* `setSingle` (localforage: ~16667, Dexie: ~23242 - 22% faster)
  * This is a method we use a lot.
* `setManySinglesBlocking` (localforage: ~no data, Dexie: ~no data)
  * This still crashed my browser at N = 100. We don't want to block anyway as row locking for Lightbeam is not a critical need when each website object is a key in our database.
*  `setManySinglesUnblocking` (localforage: ~116, Dexie: ~183 - 61% faster)
* `getRandomSet` (localforage: ~3481, Dexie: ~3850 - 11% faster)
* `getMostRecentSite`^^ (localforage: ~3892, Dexie: ~3836) - no difference
* `getLastThreeSites`^^ (localforage: ~4014, Dexie: ~3991) - no difference
* `getLastThreeDaysVisibleSites`^^ (localforage: ~3473, Dexie: ~3914 - 21% faster)
  * This is a very common joint query

^: Browser would crash on N = 10,000 websites for `getAll` and `setAll`. Writing and reading 10,000 entries from a database a hundred times is not the intended use case of this tool or Lightbeam.
^^: Now includes a `getAll` in localforage version

**Conclusion**: JSPerf tests each method against the other 50+ times and averages the results, so these results flatten out odd fluctuations, which is helpful. Dexie is still consistently faster in most cases, and especially for the methods we will use the most often.

**Comment by jonathanKingston** [comment]
Please can you handle no Dexie again, I thought this was your issue at first.
There is also the issue of the db creation being async.

**Comment by jonathanKingston** [comment]
I'm getting two errors in this branch:
"Error: missing: googleads.g.doubleclick.net"

and:
"throw new Error('setFirstParty requires a valid hostname argument');"

The first is from d3, the other from store, I'm not entirely sure how hostname would be invalid though.

This code however is invalid:
```
          await thirdParty['firstPartyHostnames']
            .forEach(async (firstPartyHostname) => {
              await this.addFirstPartyLink(firstPartyHostname, target);
            });
```

Instead you want to be doing something like this:
```
const firstPartyLinkPromises = [];
thirdParty['firstPartyHostnames'].forEach(async function (firstPartyHostname) => {
  firstPartyLinkPromises.push(this.addFirstPartyLink(firstPartyHostname, target));
});
await Promise.all(firstPartyLinkPromises);
```
(async arrows don't work yet last time I checked in Nightly).

**Comment by biancadanforth** [comment]
> I'm getting two errors in this branch

I think you're getting those errors because my most recent PR was intended to show you a bug related to failure to write the first party object into storage before the next call to `store.addFirstPartyLink` is made. This means some of the third party objects will have links that are not reciprocated in the first party object.

> This code however is invalid:

Is this code invalid because async arrows don't work yet in Nightly? If not, why else is it invalid?

**Comment by biancadanforth** [comment]
> Please can you handle no Dexie again, I thought this was your issue at first.

Are you saying you're getting an error that Dexie doesn't exist? You have to do an `npm install` (and then uninstall husky) to get it to recognize the new node module for Dexie.

> There is also the issue of the db creation being async.

Are you referring to the `store.makeNewDatabase` method? That method is not async... Can you clarify?

**Comment by jonathanKingston** [comment]
> Is this code invalid because async arrows don't work yet in Nightly? If not, why else is it invalid?

Well that and you aren't blocking, forEach won't block like you expect and using Promise.all will be quicker too.

**Comment by jonathanKingston** [comment]
> Are you saying you're getting an error that Dexie doesn't exist? You have to do an npm install (and then uninstall husky) to get it to recognize the new node module for Dexie.

I wasn't really getting an error at all on the Lightbeam page as it was being swallowed by webext's.

**Comment by jonathanKingston** [comment]
> Are you referring to the store.makeNewDatabase method? That method is not async... Can you clarify?

Maybe it isn't, not sure. This was failing other than missing Dexie. The error was that db was null when capture was calling it.
